server/src/main/java/org/opensearch/index/mapper/BinaryFieldMapper.java

```diff
@@ -240,30 +240,28 @@ protected String contentType() {
      */
     public static class CustomBinaryDocValuesField extends CustomDocValuesField {
 
-        private final ObjectArrayList<byte[]> bytesList;
-
-        private int totalSize = 0;
+        private final ArrayList<byte[]> bytesList;
 
         public CustomBinaryDocValuesField(String name, byte[] bytes) {
             super(name);
-            bytesList = new ObjectArrayList<>();
+            bytesList = new ArrayList<>();
             add(bytes);
         }
 
         public void add(byte[] bytes) {
             bytesList.add(bytes);
-            totalSize += bytes.length;
         }
 
         @Override
         public BytesRef binaryValue() {
             try {
-                CollectionUtils.sortAndDedup(bytesList);
-                int size = bytesList.size();
-                BytesStreamOutput out = new BytesStreamOutput(totalSize + (size + 1) * 5);
-                out.writeVInt(size);  // write total number of values
-                for (int i = 0; i < size; i++) {
-                    final byte[] value = bytesList.get(i);
+                // sort and dedup in place
+                CollectionUtils.sortAndDedup(bytesList, Arrays::compareUnsigned);
```

Comment:
Can we go all the way and get rid of the `sortAndDedup` method by making `bytesList` a `new TreeSet<>(Arrays::compareUnsigned)` so that it would keep the list sorted and deduped with each addition as opposed to modifying it in place here?

Reply:
I think that should be fine! TreeSet doesn't clone the objects, so we shouldn't run into heap issues. And it's Log complexity so I think that's a fine idea.